python/pyarrow/dataset.py

```diff
@@ -52,3 +52,59 @@
     SchemaPartitionScheme,
     TreeDataSource,
 )
+
+
+def open_dataset(path_or_paths, filesystem=None, partition_scheme=None,
+                 format="parquet"):
+    """
+    Open a multi-file dataset.
+
+    Parameters
+    ----------
+    path_or_paths : str or list of str
+        String path to a directory containing the data files.
+    filesystem : FileSystem, default None
+        By default will be inferred from the path.
+    partition_scheme : PartitionScheme or Schema
+        If a Schema is passed, it is interpreted as a HivePartitionScheme.
+    format : str
+        Currently only "parquet" is supported.
+
+    Returns
+    -------
+    Dataset
+
+    """
+    import pyarrow
+    from pyarrow.fs import LocalFileSystem, FileType, FileSelector
+
+    if filesystem is None:
+        # TODO handle other file systems
+        filesystem = LocalFileSystem()
+
+    if isinstance(path_or_paths, str):
+        path = filesystem.get_target_stats([path_or_paths])[0]
+        if path.type == FileType.Directory:
```

Comment:
How about adding this functionality to FileSystemDataSourceDiscovery paths_or_selector argument handling?

Reply:
> How about adding this functionality to FileSystemDataSourceDiscovery paths_or_selector argument handling?

Sounds good